### ISSUE TITLE
Firefox crashes at startup when enabling Magisk Hide

### ISSUE THREAD

**Issue by machitgarha** [topjohnwu/Magisk#2703]
Hi. A special thanks to all contributors of Magisk, especially its creator. :)

## Description

For being able to use an app, I've installed Magisk on my phone and enabled Magisk Hide. However, when I enable it, I'm unable to open Firefox; it crashes on startup every time, until disabling Magisk Hide and rebooting. Reinstalling Firefox and wiping caches don't change anything, unfortunately.

### Similar Issue(s)

#2488: I don't really know why it has been closed, but maybe because of incomplete details on the issue.

## Steps to Reproduce

1. Flash Magisk through custom recovery (e.g. TWRP).
1. Install Firefox.
1. Open Firefox. It should open normally without any problems (if Magisk Hide is disabled after the previous reboot).
1. From Magisk Manager app's settings, enable Magisk Hide.
1. Open Firefox. It crashes every time you open it.

### Steps to Fix

1. In Magisk Manager's settings, disable Magisk Hide.
1. Reboot your phone.
1. Open Firefox. It should open without crashing.

## System Information

Phone model: Samsung Galaxy Core Prime (i.e. `coreprimeve`)
ROM: ViperOS (Android 7.1.2)
Tested Magisk version(s): 20.4, 17.3
Firefox version: 68.7.0

What should I do now? I don't want to enable and disable Magisk Hide every now and then.

Thanks.

**Comment by linuxandria** [comment]
Unable to reproduce. At any rate Firefox shouldn't be affected by Hide as it does not root checks

**Comment by Didgeridoohan** [comment]
Sounds like a Webview issue...
https://didgeridoohan.com/magisk/Magisk#hn_Apps_are_force_closing

**Comment by machitgarha** [comment]
@linuxandria So, maybe this is specific to my ROM or my phone, or some apps are old here; as the link posted above suggests.

@Didgeridoohan It should not be a WebView problem, as ROM's default browser and Chrome both works without any problems. However, I'm going to take a logcat (as this is my first try on doing so, it might take some time to post it here).

**Comment by linuxandria** [comment]
> Sounds like a Webview issue...
> https://didgeridoohan.com/magisk/Magisk#hn_Apps_are_force_closing

Afaik Firefox doesn't rely on a webview. Although thanks for the link lol, one of my users was having issue with a hidden bank app fc'ing with the bromite module

**Comment by linuxandria** [comment]
> > Sounds like a Webview issue...
> > https://didgeridoohan.com/magisk/Magisk#hn_Apps_are_force_closing
> 
> Afaik Firefox doesn't rely on a webview. Although thanks for the link lol, one of my users was having issue with a hidden bank app fc'ing with the bromite module

Was actually related to the overlay included tho. The app was trying to load it for some reason
Anyhow that's neither here nor there

**Comment by machitgarha** [comment]
Where should I save the logcat?

**Comment by machitgarha** [comment]
@linuxandria @Didgeridoohan Captured the logcat: [logcat.log](https://github.com/topjohnwu/Magisk/files/4499078/logcat.log)

**Comment by topjohnwu** [comment]
Read the log, it is just a native code crash without any details. Most likely it is you running some modules that replaces some system files that Firefox expects, or simply because some weird bugs in your ROM you're using. If nothing helps, then run with MagiskHide disabled.

**Comment by machitgarha** [comment]
@topjohnwu I need MagiskHide, so I won't disable it. I'm going to use another web browser (maybe Chrome). I may try to reproduce this crash on another ROM with my phone and report it here.
BTW, thanks.

**Comment by linuxandria** [comment]
> @linuxandria @Didgeridoohan Captured the logcat: [logcat.log](https://github.com/topjohnwu/Magisk/files/4499078/logcat.log)
> 
> 

You didn't happen to restore ff from a backup did you?

**Comment by machitgarha** [comment]
@linuxandria You mean installing from a previous APK file installed in one of my previous ROMs? If that is the case, yes. However, I should note that I removed that version with all of its data associated with it and installed a clean version of Firefox from Google Play Store. Shouldn't the first installation cause an issue, am I right?

**Comment by linuxandria** [comment]
> @linuxandria You mean installing from a previous APK file installed in one of my previous ROMs? If that is the case, yes. However, I should note that I removed that version with all of its data associated with it and installed a clean version of Firefox from Google Play Store. Shouldn't the first installation cause an issue, am I right?

I find Firefox always crashes after restoring from a backup something to do with the database for logins

You're not running nightly are you?

**Comment by machitgarha** [comment]
@linuxandria No, I'm using the stable.
So the reason of the crash is the restored version, maybe. I will try a clean install at some point of my ROM and also a clean install of Firefox and report it here (I don't think in the near future, however).
Thanks for you help!

**Comment by machitgarha** [comment]
@linuxandria 

Installed Firefox from Google Play on a clean ROM with Magisk installed. When I run it, it crashes. The important part is here: Magisk Hide is disabled. This issue should be reopened, thought.

**Comment by machitgarha** [comment]
> @linuxandria
> 
> Installed Firefox from Google Play on a clean ROM with Magisk installed. When I run it, it crashes. The important part is here: Magisk Hide is disabled. This issue should be reopened, thought.

Removing Magisk with the uninstaller zip helped. Now Firefox opens without any crashes on startup. However, this is not I really want. I want to have Magisk installed.

**Comment by machitgarha** [comment]
> Read the log, it is just a native code crash without any details. Most likely it is you running some modules that replaces some system files that Firefox expects, or simply because some weird bugs in your ROM you're using. If nothing helps, then run with MagiskHide disabled.

I'm repeating my words written in two comments above somehow. However, they are important to be mentioned, and I hope this issue gets more attention.

On a clean install of my ROM, and a clean install of Magisk, (1) there are no modules installed, (2) Magisk Hide is disabled (even I didn't opened Magisk app for one time). When I open Firefox, it crashes.

The important thing is here: When I uninstall Magisk with the official uninstaller, Firefox opens without any problems, and works. This approves that the problem is not with the ROM itself, doesn't it?

As a result, I think the problem is either with Firefox or more likely with Magisk. Any help will be appreciated.

**Comment by machitgarha** [comment]
@linuxandria @topjohnwu 

On a clean install of my ROM, a clean install of Magisk, and a clean install of Firefox from Google Play Store (i.e. no restored or backup version), with no Magisk modules installed.

1. Magisk Hide disabled: When I open Firefox, it's OK.
2. Magisk Hide enabled: Without even touching the apps inside Magisk Hide, Firefox crashes on startup.

## Fixes

### Solution 1: Disable Magisk Hide

1. In Magisk Manager's settings, disable Magisk Hide.
1. Reboot your phone.
1. Open Firefox. It should open without crashing.

### Solution 2: Uninstall Magisk

1. Reboot your phone to recovery.
1. Flash Magisk uninstaller.
1. Reboot to system.
1. Open Firefox. It should open without crashing.

However, I want to neither disable Magisk Hide nor uninstall Magisk.

**Comment by machitgarha** [comment]
> Read the log, it is just a native code crash without any details. Most likely it is you running some modules that replaces some system files that Firefox expects, or simply because some weird bugs in your ROM you're using. If nothing helps, then run with MagiskHide disabled.

I don't use any modules. If an issue cannot be fixed because of whatever reason, it should not be marked as close, IMHO.

**P.S.:** Thanks for reviewing the log.

**Comment by machitgarha** [comment]
Reproduced the crash in 6 different ROMs. Here are the details (all logcats are attached):

Phone model: Samsung Galaxy Core Prime
Android version (all of them): 7.1 (Nougat) (7.1.1 or 7.1.2)
Tested Magisk version: 20.4
Firefox for Android version:

### Logs

1. AOKP ROM
    - [aokp-magisk-20.4-firefox-startup.log](https://github.com/topjohnwu/Magisk/files/4689604/aokp-magisk-20.4-firefox-startup.log)
1. Hexagon ROM:
    - [hexagon-magisk-20.4-firefox-startup.log](https://github.com/topjohnwu/Magisk/files/4689610/hexagon-magisk-20.4-firefox-startup.log)
1. LineageOS (old build)
    - [lineage-os-old-magisk-20.4-firefox-startup.log](https://github.com/topjohnwu/Magisk/files/4689616/lineage-os-old-magisk-20.4-firefox-startup.log)
1. LineageOS (new build)
    - [lineage-os-magisk-20.4-firefox-startup.log](https://github.com/topjohnwu/Magisk/files/4689620/lineage-os-magisk-20.4-firefox-startup.log)
1. RR (Resurrection Remix)
    - [resurrection-remix-magisk-20.4-firefox-startup.log](https://github.com/topjohnwu/Magisk/files/4689624/resurrection-remix-magisk-20.4-firefox-startup.log)
1. ViperOS
    - [viper-os-magisk-20.4-firefox-startup.log](https://github.com/topjohnwu/Magisk/files/4689622/viper-os-magisk-20.4-firefox-startup.log)

**Comment by machitgarha** [comment]
@topjohnwu

> Read the log, it is just a native code crash without any details. Most likely it is you running some modules that replaces some system files that Firefox expects, or simply because some weird bugs in your ROM you're using. If nothing helps, then run with MagiskHide disabled.

As a result of my two comments above, the reason is not the ROM. As I didn't installed any modules while testing, the reason is not the Magisk modules. So the problem is either with Firefox or with Magisk Hide. The problem should specifically be related to Firefox Gecko, and maybe there is a conflict between Magisk Hide stuff and that Firefox component.

So, I think this issue should be reopened.

**Comment by linuxandria** [comment]
Thing is I can't reproduce on any Firefox version, although I am testing on Android 10. So it's definitely not magisk related and probably not directly Firefox's fault

**Comment by linuxandria** [comment]
You'd be better off opening a detailed issue on BugZilla

**Comment by machitgarha** [comment]
> Thing is I can't reproduce on any Firefox version, although I am testing on Android 10. So it's definitely not magisk related and probably not directly Firefox's fault

Frustrating. However, there is another issue that produced the same crash (as I mentioned as a duplicate issue). It seems it is only reproducable on Android 7.1.* versions.

> You'd be better off opening a detailed issue on BugZilla

You mean Mozilla's BugZilla?

**Comment by machitgarha** [comment]
After testing Firefox Nightly versions, the crash doesn't happen there. I tested 5 Firefox distributions, and, after enabling Magisk Hide, the result is (without Magisk Hide being enabled, there is no crash, obviously):

- Crashes on startup:
    - Firefox stable (68.8.1)
    - Firefox beta (75.0.0-beta.6)
- Crashes on shutdown:
    - Firefox Preview (76.0)
- Works without crashes:
    - Firefox Nightly (78.0a1)
    - Firefox Preview Nightly (78.0a1)

As a result, the problem seems to be with the Firefox itself (maybe with specific Android 7.1 ROMs), and as it seems, it has been fixed in recent versions.

So, thank you all, I leave this issue closed and I think it should be marked as fixed.

**P.S.**: A special thanks to you, @linuxandria, for your great help!
